### PR TITLE
Set up YCSB benchmark

### PR DESCRIPTION
**Pull request by jeffa5**
This ended up being a bit larger due to a lot of refactoring to handle different benchmark methods but should be more extensible in future for others.

Probably some follow-up things to do after merging:
- Add ycsb to CI
- setup python files for ci configuration and an all configuration rather than relying on flags everywhere

### CODE REVIEW

File: benchmark/ycsb.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        bench = [
+            "bin/go-ycsb",
+            "run",
+            "etcd",
+            "--prop",
+            f"etcd.endpoints={self.config.scheme()}://127.0.0.1:{self.config.port}",
+            "--prop",
+            "measurementtype=raw",
+            "--prop",
+            f"measurement.raw.output_file={timings_file}",
```
Comment (jeffa5) [inline benchmark/ycsb.py:100]: go-ycsb doesn't currently support the raw output measurement style so will take a look at adding raising an issue on their repo.

**Comment by jeffa5** [comment]
This gets the ycsb benchmark into a state where it runs but doesn't currently output raw data to a CSV since it isn't currently implemented. I'll have a look at how difficult that would be.

We could switch to the original ycsb but we'd need to setup an etcd client which wouldn't be very nice either.

File: benchmark/ycsb.py (hunk 1 above)
Comment (jeffa5) [reply benchmark/ycsb.py:92]: https://github.com/pingcap/go-ycsb/issues/255

**Comment by heidihoward** [comment]
Just so I understand, you can get the processed results from the go-ycsb client printed to stdout but you'd like to have the raw results in an easy to process format for doing your own analysis?

**Comment by heidihoward** [comment]
I think it would be preferable to use go-ycsb over writing our own client, though I think we will eventually need to do the latter as well

**Comment by jeffa5** [comment]
> Just so I understand, you can get the processed results from the go-ycsb client printed to stdout but you'd like to have the raw results in an easy to process format for doing your own analysis?

Yes.

**Pull request merged**